APPLE: Enable Building as Framework

dgovil opened the pull request:
### Description of Change(s)

This PR separates out just the framework build part of https://github.com/PixarAnimationStudios/OpenUSD/pull/2969 into an isolated PR, and addresses the notes from @meshula in the process.

**Overview of Changes**

* `build_usd.py` now adds a `--build-apple-framework` and `--no-build-apple-framework` to enable building frameworks. Frameworks are enabled by default for embedded OS builds.
* When building as a framework, builds are monolithic.
* Framework builds force TBB and OpenSubdiv to be statically linked. This can be revised in the future, but we discussed this previously, and decided it was the lesser of two evils given we're moving to GCD support.
* The CMake build calls out to a zsh script (zsh is guaranteed to exist on macOS) to build the framework. This is used because CMake's own native framework build support would be far more invasive a change currently. 
* A new configuration option `PXR_EXTRA_BINARY_RELATIVE_RESOURCES` allows the resource directory to be configured relative to the binary. This avoids hardcoding patterns in the CPP files, and allows configuration in CMake. This was requested by @meshula  as well.

### Checklist

- [X] I have created this PR based on the dev branch

- [X] I have followed the [coding conventions](https://openusd.org/release/api/_page__coding__guidelines.html)

- [ ] I have added unit tests that exercise this functionality (Reference: 
  [testing guidelines](https://openusd.org/release/api/_page__testing__guidelines.html))

- [X] I have verified that all unit tests pass with the proposed changes

- [X] I have submitted a signed Contributor License Agreement (Reference: 
  [Contributor License Agreement instructions](https://openusd.org/release/contributing_to_usd.html#contributor-license-agreement))

dgovil committed:
Configure CMake to setup the Framework arguments

dgovil committed:
Copy back Framework build script

dgovil committed:
Modify search paths for frameworks

dgovil committed:
Remove extra option

jesschimein commented:
Filed as internal issue #USD-11201

(This is an automated message. See [here](https://openusd.org/release/contributing_to_usd.html#github-issues) for more information.)

tlewis commented on cmake/defaults/CXXDefaults.cmake:
```diff
@@ -86,3 +86,12 @@ if (PXR_PREFER_SAFETY_OVER_SPEED)
+
+# Set that Apple Framework is being build
+if (PXR_BUILD_APPLE_FRAMEWORK)
+    _add_define("PXR_BUILD_APPLE_FRAMEWORK")
```
Hi @dgovil - I'm curious what the motivation is for this. It's currently not used by anything so is this ok to remove?

dgovil replied:
Yeah I think it can be removed. It was vestigial from when I spilt up the PR

tlewis commented on cmake/defaults/Options.cmake:
```diff
@@ -165,7 +173,21 @@ set(PXR_EXTRA_PLUGINS ""
+set(PXR_EXTRA_BINARY_RELATIVE_RESOURCES ""
+        CACHE
+        STRING
+        "An optional resources folder relative to the binary to look for resources.")
```
@dgovil , is this option needed or can its functionality be accomplished by using `PXR_INSTALL_LOCATION`?

dgovil replied:
This one is unfortunately needed because of the way the path searching assumptions that are in base/plug/initConfig.cpp and usd/usdMtlx/utils.cpp

In a framework, the install path I believe is where the framework gets installed, but the Resources folder is within the folder. But the folder structure can have symlinks etc that require the pathing to be resolved relative to the runtime binary location

tlewis commented on cmake/resources/AppleFrameworkBuild.zsh.in:
```diff
@@ -0,0 +1,87 @@
+ditto "${PROJECT_BINARY_DIR}/Info.plist" "${FRAMEWORK_PLIST_LOCATION}"
+
+
+# Copy the MaterialX libraries if they exist
```
@dgovil this MaterialX library copying seems unnecessary as `usdMtlx` already copies these files as long as `MATERIALX_STDLIB_DIR` is defined (which seems to be the case if `MaterialXConfig.cmake` is ingested by the OpenUSD build).

I removed this extra copy in a local test framework build and the MaterialX functionality seemed to still work fine, even if I removed the original MaterialX install path which gets baked into the usdMtlx lib. The relative path discovery of the mtlx files seems to work fine from within the .framework.

dgovil commented:
Hmm interesting. Maybe something else has changed in the past year to make
it no longer necessary.
I guess we can avoid it for now and then add it back if the need comes up
again.

Out of curiosity, did you check if the hydra plugins are found?

On Wed, 1 Jul 2026 at 17:51, tlewis ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cmake/resources/AppleFrameworkBuild.zsh.in
> <https://github.com/PixarAnimationStudios/OpenUSD/pull/3713#discussion_r3509819219>
> :
>
> > +  FRAMEWORK_LIB_PATH="${FRAMEWORK_DIR}/Versions/A/${FRAMEWORK_NAME}"
> +  ***@***.***/${FRAMEWORK_NAME}.framework/${FRAMEWORK_NAME}"
> +fi
> +
> +echo "Creating Resources Root: ${FRAMEWORK_RESOURCES_DIR}"
> +mkdir -p ${FRAMEWORK_RESOURCES_DIR}
> +
> +echo "Creating Headers Root: ${FRAMEWORK_HEADERS_DIR}"
> +mkdir -p ${FRAMEWORK_HEADERS_DIR}
> +
> +# Copy the plist over
> +echo "Copying files into ${FRAMEWORK_DIR}"
> +ditto "${PROJECT_BINARY_DIR}/Info.plist" "${FRAMEWORK_PLIST_LOCATION}"
> +
> +
> +# Copy the MaterialX libraries if they exist
>
> @dgovil <https://github.com/dgovil> this MaterialX library copying seems
> unnecessary as usdMtlx already copies these files as long as
> MATERIALX_STDLIB_DIR is defined (which seems to be the case if
> MaterialXConfig.cmake is ingested by the OpenUSD build).
>
> I removed this extra copy in a local test framework build and the
> MaterialX functionality seemed to still work fine, even if I removed the
> original MaterialX install path which gets baked into the usdMtlx lib. The
> relative path discovery of the mtlx files seems to work fine from within
> the .framework.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/OpenUSD/pull/3713?email_source=notifications&email_token=ABB4XUWS65XHILW67X46SR35CWW2DA5CNFSNUABKM5UWIORPF5TWS5BNNB2WEL2QOVWGYUTFOF2WK43UKJSXM2LFO4XTINRRGM3TEOJUHE3KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#pullrequestreview-4613729496>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB4XUU4UW4PTJXSDW372BL5CWW2DAVCNFSNUABEKJSXA33TNF2G64TZHM2TQMJWHAYTIMZ3JFZXG5LFHMZTEMJTGY2TAOJSHCQXMAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
